File: LINEBOT/admin_bot/app/services/booking_service.py
import time
import re
from sqlalchemy.orm import Session
from ..models import Booking
from .schedule_service import update_schedule_slot_status


def get_all_bookings(db: Session) -> list[Booking]:
    return db.query(Booking).order_by(Booking.id.desc()).all()


def create_booking(db: Session, data: dict) -> Booking:
    now_unix = int(time.time())
    norm_date = data["date"].strip()
    norm_event = data["eventName"].strip()
    norm_slot = re.sub(r'\s+', '', data["timeSlot"])
    norm_cam = (data.get("cameraType") or "").strip()

    existing_bookings = db.query(Booking).filter(Booking.date == norm_date).all()
    for b in existing_bookings:
        if b.status == "cancelled":
            continue
        same_event = not norm_event or not b.event_name or b.event_name.strip() == norm_event
        same_slot = re.sub(r'\s+', '', b.time_slot or "") == norm_slot
        b_cam = (b.camera_type or "").strip()
        same_cam = not norm_cam or not b_cam or norm_cam.lower() == b_cam.lower()

        if same_event and same_slot and same_cam:
            raise ValueError(f"รอบเวลา {data['timeSlot']} น. สำหรับกล้อง {norm_cam or 'รุ่นนี้'} มีผู้ทำรายการจองคิวไว้แล้ว")

    new_booking = Booking(
        date=norm_date,
        event_name=norm_event,
        time_slot=data["timeSlot"].strip(),
        customer_name=data["customerName"].strip(),
        customer_phone=data["customerPhone"].strip(),
        line_display_name=data.get("lineDisplayName"),
        line_user_id=data.get("lineUserId"),
        camera_type=data.get("cameraType"),
        status=data.get("status", "pending"),
        payment_status=data.get("paymentStatus", "unpaid"),
        deposit_amount=data.get("depositAmount", 0),
        remaining_amount=data.get("remainingAmount", 0),
        notes=data.get("notes"),
        created_at=now_unix,
    )
    db.add(new_booking)
    db.commit()
    db.refresh(new_booking)

    if new_booking.status != "cancelled":
        try:
            update_schedule_slot_status(
                db=db,
                target_date=norm_date,
                target_event_name=norm_event,
                time_slot=data["timeSlot"],
                new_status="booked",
                camera_type=data.get("cameraType"),
            )
        except Exception as e:
            print(f"Warning: could not auto-sync slot status to schedules: {e}")

    return new_booking


def update_booking_status(
    db: Session,
    booking_id: int,
    status: str | None = None,
    notes: str | None = None,
    payment_status: str | None = None,
    deposit_amount: int | None = None,
    remaining_amount: int | None = None,
) -> bool:
    booking = db.query(Booking).filter(Booking.id == booking_id).first()
    if not booking:
        return False

    if status is not None:
        booking.status = status
    if notes is not None:
        booking.notes = notes
    if payment_status is not None:
        booking.payment_status = payment_status
    if deposit_amount is not None:
        booking.deposit_amount = deposit_amount
    if remaining_amount is not None:
        booking.remaining_amount = remaining_amount

    db.commit()

    if status == "cancelled":
        try:
            other_bookings = db.query(Booking).filter(Booking.date == booking.date).all()
            clean_slot = re.sub(r'\s+', '', booking.time_slot or "")
            existing_cam_norm = (booking.camera_type or "").strip().lower()

            has_other_active = any(
                b.id != booking_id and b.status != "cancelled" and
                re.sub(r'\s+', '', b.time_slot or "") == clean_slot and
                (not existing_cam_norm or not (b.camera_type or "").strip() or (b.camera_type or "").strip().lower() == existing_cam_norm)
                for b in other_bookings
            )

            if not has_other_active:
                update_schedule_slot_status(
                    db=db,
                    target_date=booking.date,
                    target_event_name=booking.event_name,
                    time_slot=booking.time_slot,
                    new_status="available",
                    camera_type=booking.camera_type or None,
                )
        except Exception as e:
            print(f"Warning: could not sync slot on cancellation: {e}")
    elif status in ["confirmed", "pending"]:
        try:
            update_schedule_slot_status(
                db=db,
                target_date=booking.date,
                target_event_name=booking.event_name,
                time_slot=booking.time_slot,
                new_status="booked",
                camera_type=booking.camera_type or None,
            )
        except Exception as e:
            print(f"Warning: could not sync slot on update: {e}")

    return True


def confirm_or_create_booking(db: Session, data: dict) -> Booking:
    booking_id = data.get("bookingId")
    norm_date = data["date"].strip()
    norm_event = data["eventName"].strip()
    norm_slot = re.sub(r'\s+', '', data["timeSlot"])
    norm_cam = (data.get("cameraType") or "").strip().lower()

    if booking_id:
        existing = db.query(Booking).filter(Booking.id == booking_id).first()
        if existing:
            existing.status = data.get("status", "confirmed")
            existing.payment_status = data.get("paymentStatus", "paid")
            if data.get("depositAmount") is not None:
                existing.deposit_amount = data["depositAmount"]
            if data.get("remainingAmount") is not None:
                existing.remaining_amount = data["remainingAmount"]
            if data.get("notes") is not None:
                existing.notes = data["notes"]
            db.commit()
            db.refresh(existing)

            if existing.status != "cancelled":
                try:
                    update_schedule_slot_status(
                        db=db,
                        target_date=norm_date,
                        target_event_name=norm_event,
                        time_slot=data["timeSlot"],
                        new_status="booked",
                        camera_type=data.get("cameraType"),
                    )
                except Exception as e:
                    print(f"Warning: could not sync slot on confirm: {e}")
            return existing

    # Check for pending booking
    existing_bookings = db.query(Booking).filter(Booking.date == norm_date).all()
    pending = None
    for b in existing_bookings:
        if b.status != "pending":
            continue
        same_event = not norm_event or not b.event_name or b.event_name.strip() == norm_event
        same_slot = re.sub(r'\s+', '', b.time_slot or "") == norm_slot
        b_cam = (b.camera_type or "").strip().lower()
        same_cam = not norm_cam or not b_cam or norm_cam == b_cam
        if same_event and same_slot and same_cam:
            pending = b
            break

    if pending:
        pending.status = data.get("status", "confirmed")
        pending.payment_status = data.get("paymentStatus", "paid")
        if data.get("depositAmount") is not None:
            pending.deposit_amount = data["depositAmount"]
        if data.get("remainingAmount") is not None:
            pending.remaining_amount = data["remainingAmount"]
        if data.get("notes") is not None:
            pending.notes = data["notes"]
        pending.customer_name = data["customerName"].strip()
        pending.customer_phone = data["customerPhone"].strip()
        if data.get("lineDisplayName"):
            pending.line_display_name = data["lineDisplayName"]
        if data.get("lineUserId"):
            pending.line_user_id = data["lineUserId"]
        db.commit()
        db.refresh(pending)

        if pending.status != "cancelled":
            try:
                update_schedule_slot_status(
                    db=db,
                    target_date=norm_date,
                    target_event_name=norm_event,
                    time_slot=data["timeSlot"],
                    new_status="booked",
                    camera_type=data.get("cameraType"),
                )
            except Exception as e:
                print(f"Warning: could not sync slot on confirm: {e}")
        return pending

    return create_booking(db, data)


def cancel_pending_booking_if_exists(db: Session, draft: dict) -> bool:
    booking_id = draft.get("bookingId")
    if booking_id:
        return update_booking_status(db, booking_id, status="cancelled")

    norm_date = draft["date"].strip()
    norm_event = draft["eventName"].strip()
    norm_slot = re.sub(r'\s+', '', draft["timeSlot"])
    norm_cam = (draft.get("cameraType") or "").strip().lower()

    existing_bookings = db.query(Booking).filter(Booking.date == norm_date).all()
    for b in existing_bookings:
        if b.status != "pending":
            continue
        same_event = not norm_event or not b.event_name or b.event_name.strip() == norm_event
        same_slot = re.sub(r'\s+', '', b.time_slot or "") == norm_slot
        b_cam = (b.camera_type or "").strip().lower()
        same_cam = not norm_cam or not b_cam or norm_cam == b_cam
        if same_event and same_slot and same_cam:
            return update_booking_status(db, b.id, status="cancelled")

    return False

File: src/shared/services/booking-service.ts
import { db } from '@/db/db';
import { bookings } from '@/db/schema';
import { eq, desc } from 'drizzle-orm';
import { updateScheduleSlotStatus } from '@/shared/services/schedule-service';
import { isSameCamera } from '@/shared/utils/camera-utils';

export interface BookingRecord {
  id: number;
  date: string;
  eventName: string;
  timeSlot: string;
  customerName: string;
  customerPhone: string;
  lineDisplayName: string | null;
  lineUserId: string | null;
  cameraType: string | null;
  status: 'pending' | 'confirmed' | 'cancelled';
  paymentStatus: 'unpaid' | 'deposit' | 'paid';
  depositAmount?: number | null;
  remainingAmount?: number | null;
  notes: string | null;
  createdAt: number;
}

export interface FetchBookingsOptions {
  limit?: number;
  offset?: number;
}

// In-memory cache for all bookings to speed up /admin and fetchBookingsAction
let cachedAllBookings: { data: BookingRecord[]; expiresAt: number } | null = null;
const BOOKINGS_CACHE_TTL_MS = 30 * 1000; // 30 seconds

export function invalidateBookingCache() {
  cachedAllBookings = null;
}

/**
 * Fetch all booking records ordered by newest first with optional pagination
 */
export async function getAllBookings(options?: FetchBookingsOptions): Promise<BookingRecord[]> {
  const isDefaultFetch = !options?.limit && !options?.offset;
  const now = Date.now();

  if (isDefaultFetch && cachedAllBookings && now < cachedAllBookings.expiresAt) {
    return cachedAllBookings.data;
  }

  let query = db.select().from(bookings).orderBy(desc(bookings.id));
  if (options?.limit) {
    query = query.limit(options.limit) as any;
  }
  if (options?.offset) {
    query = query.offset(options.offset) as any;
  }
  const records = await query;
  const result = records as BookingRecord[];

  if (isDefaultFetch) {
    cachedAllBookings = { data: result, expiresAt: now + BOOKINGS_CACHE_TTL_MS };
  }

  return result;
}

/**
 * Create a new booking (from LINE Chatbot or Admin UI)
 */
export async function createBooking(data: {
  date: string;
  eventName: string;
  timeSlot: string;
  customerName: string;
  customerPhone: string;
  lineDisplayName?: string;
  lineUserId?: string;
  cameraType?: string;
  status?: 'pending' | 'confirmed' | 'cancelled';
  paymentStatus?: 'unpaid' | 'deposit' | 'paid';
  depositAmount?: number;
  remainingAmount?: number;
  notes?: string;
}): Promise<BookingRecord> {
  const nowUnix = Math.floor(Date.now() / 1000);
  const normDate = data.date.trim();
  const normEvent = data.eventName.trim();
  const normSlot = data.timeSlot.replace(/\s+/g, '');
  const normCam = (data.cameraType || '').trim();

  // Check for duplicate active booking for the same date, event, slot, and camera
  const existingBookings = await db
    .select()
    .from(bookings)
    .where(eq(bookings.date, normDate));

  const isDuplicate = existingBookings.some(b => {
    if (b.status === 'cancelled') return false;

    const sameEvent = !normEvent || !b.eventName || (b.eventName || '').trim() === normEvent;
    const sameSlot = (b.timeSlot || '').replace(/\s+/g, '') === normSlot;
    const sameCam = isSameCamera(b.cameraType, data.cameraType);

    return sameEvent && sameSlot && sameCam;
  });

  if (isDuplicate) {
    throw new Error(`รอบเวลา ${data.timeSlot} น. สำหรับกล้อง ${normCam || 'รุ่นนี้'} มีผู้ทำรายการจองคิวไว้แล้ว`);
  }

  const [inserted] = await db
    .insert(bookings)
    .values({
      date: normDate,
      eventName: normEvent,
      timeSlot: data.timeSlot.trim(),
      customerName: data.customerName.trim(),
      customerPhone: data.customerPhone.trim(),
      lineDisplayName: data.lineDisplayName || null,
      lineUserId: data.lineUserId || null,
      cameraType: data.cameraType || null,
      status: data.status || 'pending',
      paymentStatus: data.paymentStatus || 'unpaid',
      depositAmount: data.depositAmount ?? 0,
      remainingAmount: data.remainingAmount ?? 0,
      notes: data.notes || null,
      createdAt: nowUnix,
    })
    .returning();

  if (inserted.status !== 'cancelled') {
    try {
      await updateScheduleSlotStatus(normDate, normEvent, data.timeSlot, 'booked', data.cameraType || undefined);
    } catch (e) {
      console.warn('Could not auto-sync slot status to schedules table:', e);
    }
  }

  invalidateBookingCache();
  return inserted as BookingRecord;
}

/**
 * Update booking status ('pending', 'confirmed', 'cancelled')
 */
export async function updateBookingStatus(
  id: number,
  status?: 'pending' | 'confirmed' | 'cancelled',
  notes?: string,
  paymentStatus?: 'unpaid' | 'deposit' | 'paid',
  depositAmount?: number,
  remainingAmount?: number
): Promise<boolean> {
  const [existing] = await db.select().from(bookings).where(eq(bookings.id, id));
  if (!existing) return false;

  await db
    .update(bookings)
    .set({
      ...(status !== undefined && { status }),
      ...(notes !== undefined && { notes }),
      ...(paymentStatus !== undefined && { paymentStatus }),
      ...(depositAmount !== undefined && { depositAmount }),
      ...(remainingAmount !== undefined && { remainingAmount }),
    })
    .where(eq(bookings.id, id));

  // If status is updated
  if (status === 'cancelled') {
    try {
      // Check if there are other active bookings for the same date, slot, and camera
      const otherBookings = await db
        .select()
        .from(bookings)
        .where(eq(bookings.date, existing.date));

      const cleanSlot = (existing.timeSlot || '').replace(/\s+/g, '');
      const hasOtherActiveBookingForSameCam = otherBookings.some(
        b => b.id !== id && b.status !== 'cancelled' &&
             (b.timeSlot || '').replace(/\s+/g, '') === cleanSlot &&
             isSameCamera(b.cameraType, existing.cameraType)
      );

      if (!hasOtherActiveBookingForSameCam) {
        await updateScheduleSlotStatus(existing.date, existing.eventName, existing.timeSlot, 'available', existing.cameraType || undefined);
      }
    } catch (e) {
      console.warn('Could not sync slot status on cancellation:', e);
    }
  } else if (status === 'confirmed' || status === 'pending') {
    try {
      await updateScheduleSlotStatus(existing.date, existing.eventName, existing.timeSlot, 'booked', existing.cameraType || undefined);
    } catch (e) {
      console.warn('Could not sync slot status on update:', e);
    }
  }

  invalidateBookingCache();
  return true;
}

export interface UpdateBookingInput {
  date: string;
  eventName: string;
  timeSlot: string;
  customerName: string;
  customerPhone: string;
  lineDisplayName?: string | null;
  cameraType?: string | null;
  status: 'pending' | 'confirmed' | 'cancelled';
  paymentStatus: 'unpaid' | 'deposit' | 'paid';
  depositAmount?: number | null;
  remainingAmount?: number | null;
  notes?: string | null;
}

/**
 * Full update of all fields in a booking record (Admin Edit)
 * Automatically synchronizes slot availability between old and new slot/camera/date
 */
export async function updateBookingFull(
  id: number,
  data: UpdateBookingInput
): Promise<BookingRecord> {
  const [existing] = await db.select().from(bookings).where(eq(bookings.id, id));
  if (!existing) {
    throw new Error('ไม่พบรายการจองที่ต้องการแก้ไข');
  }

  const normDate = data.date.trim();
  const normEvent = data.eventName.trim();
  const normSlot = data.timeSlot.trim();
  const cleanNewSlot = normSlot.replace(/\s+/g, '');
  const cleanOldSlot = (existing.timeSlot || '').replace(/\s+/g, '');
  const oldCam = (existing.cameraType || '').trim();
  const newCam = (data.cameraType || '').trim();

  const slotOrCamChanged =
    existing.date.trim() !== normDate ||
    (existing.eventName || '').trim() !== normEvent ||
    cleanOldSlot !== cleanNewSlot ||
    !isSameCamera(oldCam, newCam);

  const statusChanged = existing.status !== data.status;

  const [updated] = await db
    .update(bookings)
    .set({
      date: normDate,
      eventName: normEvent,
      timeSlot: normSlot,
      customerName: data.customerName.trim(),
      customerPhone: data.customerPhone.trim(),
      lineDisplayName: data.lineDisplayName ? data.lineDisplayName.trim() : null,
      cameraType: newCam || null,
      status: data.status,
      paymentStatus: data.paymentStatus,
      depositAmount: data.depositAmount ?? 0,
      remainingAmount: data.remainingAmount ?? 0,
      notes: data.notes ? data.notes.trim() : null,
    })
    .where(eq(bookings.id, id))
    .returning();

  // Handle slot sync:
  // 1. Release old slot if slot/cam/date changed OR status changed to cancelled
  const shouldReleaseOldSlot =
    (slotOrCamChanged || (statusChanged && data.status === 'cancelled')) &&
    existing.status !== 'cancelled';

  if (shouldReleaseOldSlot) {
    try {
      const otherOldBookings = await db
        .select()
        .from(bookings)
        .where(eq(bookings.date, existing.date));

      const hasOtherActiveBooking = otherOldBookings.some(
        b =>
          b.id !== id &&
          b.status !== 'cancelled' &&
          (b.timeSlot || '').replace(/\s+/g, '') === cleanOldSlot &&
          isSameCamera(b.cameraType, existing.cameraType)
      );

      if (!hasOtherActiveBooking) {
        await updateScheduleSlotStatus(
          existing.date,
          existing.eventName,
          existing.timeSlot,
          'available',
          existing.cameraType || undefined
        );
      }
    } catch (e) {
      console.warn('Could not release old slot on booking edit:', e);
    }
  }

  // 2. Mark new slot as booked if updated booking is active
  if (data.status !== 'cancelled') {
    try {
      await updateScheduleSlotStatus(
        normDate,
        normEvent,
        normSlot,
        'booked',
        newCam || undefined
      );
    } catch (e) {
      console.warn('Could not reserve new slot on booking edit:', e);
    }
  }

  invalidateBookingCache();
  return updated as BookingRecord;
}

/**
 * Confirm an existing pending booking or create a new confirmed booking
 */
export async function confirmOrCreateBooking(data: {
  bookingId?: number;
  date: string;
  eventName: string;
  timeSlot: string;
  customerName: string;
  customerPhone: string;
  lineDisplayName?: string;
  lineUserId?: string;
  cameraType?: string;
  status?: 'pending' | 'confirmed' | 'cancelled';
  paymentStatus?: 'unpaid' | 'deposit' | 'paid';
  depositAmount?: number;
  remainingAmount?: number;
  notes?: string;
}): Promise<BookingRecord> {
  const normDate = data.date.trim();
  const normEvent = data.eventName.trim();
  const normSlot = data.timeSlot.replace(/\s+/g, '');
  const normCam = (data.cameraType || '').trim().toLowerCase();

  // 1. If explicit bookingId provided, update that booking directly
  if (data.bookingId) {
    const [existing] = await db.select().from(bookings).where(eq(bookings.id, data.bookingId));
    if (existing) {
      const [updated] = await db
        .update(bookings)
        .set({
          status: data.status || 'confirmed',
          paymentStatus: data.paymentStatus || 'paid',
          depositAmount: data.depositAmount ?? existing.depositAmount ?? 0,
          remainingAmount: data.remainingAmount ?? existing.remainingAmount ?? 0,
          notes: data.notes || existing.notes,
        })
        .where(eq(bookings.id, data.bookingId))
        .returning();

      if (updated.status !== 'cancelled') {
        try {
          await updateScheduleSlotStatus(normDate, normEvent, data.timeSlot, 'booked', data.cameraType || undefined);
        } catch (e) {
          console.warn('Could not sync slot status on confirm bookingId:', e);
        }
      }

      return updated as BookingRecord;
    }
  }

  // 2. Check if there is an existing pending booking matching date, event, slot, camera
  const existingBookings = await db
    .select()
    .from(bookings)
    .where(eq(bookings.date, normDate));

  const pendingBooking = existingBookings.find(b => {
    if (b.status !== 'pending') return false;
    const sameEvent = !normEvent || !b.eventName || (b.eventName || '').trim() === normEvent;
    const sameSlot = (b.timeSlot || '').replace(/\s+/g, '') === normSlot;
    const sameCam = isSameCamera(b.cameraType, data.cameraType);
    return sameEvent && sameSlot && sameCam;
  });

  if (pendingBooking) {
    const [updated] = await db
      .update(bookings)
      .set({
        status: data.status || 'confirmed',
        paymentStatus: data.paymentStatus || 'paid',
        depositAmount: data.depositAmount ?? pendingBooking.depositAmount ?? 0,
        remainingAmount: data.remainingAmount ?? pendingBooking.remainingAmount ?? 0,
        notes: data.notes || pendingBooking.notes,
        customerName: data.customerName.trim(),
        customerPhone: data.customerPhone.trim(),
        lineDisplayName: data.lineDisplayName || pendingBooking.lineDisplayName,
        lineUserId: data.lineUserId || pendingBooking.lineUserId,
      })
      .where(eq(bookings.id, pendingBooking.id))
      .returning();

    if (updated.status !== 'cancelled') {
      try {
        await updateScheduleSlotStatus(normDate, normEvent, data.timeSlot, 'booked', data.cameraType || undefined);
      } catch (e) {
        console.warn('Could not sync slot status on confirm pendingBooking:', e);
      }
    }

    return updated as BookingRecord;
  }

  // 3. Otherwise create new booking via createBooking
  return await createBooking(data);
}


/**
 * Cancel pending booking if exists
 */
export async function cancelPendingBookingIfExists(draft: {
  bookingId?: number;
  date: string;
  eventName: string;
  timeSlot: string;
  cameraType?: string;
}): Promise<boolean> {
  if (draft.bookingId) {
    return await updateBookingStatus(draft.bookingId, 'cancelled');
  }

  const normDate = draft.date.trim();
  const normEvent = draft.eventName.trim();
  const normSlot = draft.timeSlot.replace(/\s+/g, '');
  const normCam = (draft.cameraType || '').trim().toLowerCase();

  const existingBookings = await db
    .select()
    .from(bookings)
    .where(eq(bookings.date, normDate));

  const pendingBooking = existingBookings.find(b => {
    if (b.status !== 'pending') return false;
    const sameEvent = !normEvent || !b.eventName || (b.eventName || '').trim() === normEvent;
    const sameSlot = (b.timeSlot || '').replace(/\s+/g, '') === normSlot;
    const sameCam = isSameCamera(b.cameraType, draft.cameraType);
    return sameEvent && sameSlot && sameCam;
  });

  if (pendingBooking) {
    await updateBookingStatus(pendingBooking.id, 'cancelled');
    return true;
  }
  return false;
}

/**
 * Delete a booking record
 */
export async function deleteBooking(id: number): Promise<boolean> {
  const [existing] = await db.select().from(bookings).where(eq(bookings.id, id));
  await db.delete(bookings).where(eq(bookings.id, id));

  if (existing && existing.status !== 'cancelled') {
    try {
      const otherBookings = await db
        .select()
        .from(bookings)
        .where(eq(bookings.date, existing.date));

      const cleanSlot = (existing.timeSlot || '').replace(/\s+/g, '');
      const hasOtherActiveBookingForSameCam = otherBookings.some(
        b => b.id !== id && b.status !== 'cancelled' &&
             (b.timeSlot || '').replace(/\s+/g, '') === cleanSlot &&
             isSameCamera(b.cameraType, existing.cameraType)
      );

      if (!hasOtherActiveBookingForSameCam) {
        await updateScheduleSlotStatus(existing.date, existing.eventName, existing.timeSlot, 'available', existing.cameraType || undefined);
      }
    } catch (e) {
      console.warn('Could not sync slot status on delete:', e);
    }
  }

  invalidateBookingCache();
  return true;
}
